pkg/manifest/os.go

```diff
@@ -737,6 +739,36 @@ func (p *OS) serialize() osbuild.Pipeline {
 		pipeline.AddStages(osbuild.GenFileNodesStages(p.Files)...)
 	}
 
+	// write modularity related configuration files
+	if len(p.moduleSpecs) > 0 {
+		pipeline.AddStages(osbuild.GenDNFModuleConfigStages(p.moduleSpecs)...)
+
+		var failsafeFiles []*fsnode.File
+
+		// the failsafe file is a blob of YAML returned directly from the depsolver,
+		// we write them as 'normal files' without a special stage
+		for _, module := range p.moduleSpecs {
+			moduleFailsafeFile, err := fsnode.NewFile(module.FailsafeFile.Path, nil, nil, nil, []byte(module.FailsafeFile.Data))
+
+			if err != nil {
+				panic("failed to create module failsafe file")
+			}
+
+			failsafeFiles = append(failsafeFiles, moduleFailsafeFile)
+		}
+
+		failsafeDir, err := fsnode.NewDirectory("/var/lib/dnf/modulefailsafe", nil, nil, nil, true)
+
+		if err != nil {
+			panic("failed to create module failsafe directory")
+		}
+
+		pipeline.AddStages(osbuild.GenDirectoryNodesStages([]*fsnode.Directory{failsafeDir})...)
```

Comment:
Would it make sense to make this conditional `if len(failsafeFiles) > 0` ?

Reply:
It could be; but it'd be an error to not have any `failsafeFiles` when there *were* modules in the depsolve result.

Reply:
Reading above, there's no way for `len(failsafeFiles)` to have be different from `len(moduleSpecs)`, even if one or more `module.FailsafeFile.Data` are empty (which it should never be).